clippy_lints/src/hashset_insert_after_contains.rs

```diff
@@ -0,0 +1,184 @@
+        _ => None,
+    });
+    match expr.kind {
+        ExprKind::MethodCall(
+            path,
+            receiver,
+            [
+                Expr {
+                    kind: ExprKind::AddrOf(_, _, value),
+                    span: value_span,
+                    ..
+                },
+            ],
+            _,
+        ) => {
+            let receiver_ty = cx.typeck_results().expr_ty(receiver);
+            if value_span.ctxt() == expr.span.ctxt()
+                && is_type_diagnostic_item(cx, receiver_ty, sym::HashSet)
+                && path.ident.name == sym!(contains)
+            {
+                Some(ContainsExpr { receiver, value })
+            } else {
+                None
+            }
+        },
+        _ => None,
+    }
+}
+
+struct InsertExpr<'tcx> {
+    receiver: &'tcx Expr<'tcx>,
+    value: &'tcx Expr<'tcx>,
+}
+fn try_parse_insert<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) -> Option<InsertExpr<'tcx>> {
+    if let ExprKind::MethodCall(path, receiver, [value], _) = expr.kind {
+        let receiver_ty = cx.typeck_results().expr_ty(receiver);
+        if is_type_diagnostic_item(cx, receiver_ty, sym::HashSet) && path.ident.name == sym!(insert) {
+            Some(InsertExpr { receiver, value })
+        } else {
+            None
+        }
+    } else {
+        None
+    }
+}
+
+fn find_insert_calls<'tcx>(cx: &LateContext<'tcx>, contains_expr: &ContainsExpr<'tcx>, expr: &'tcx Expr<'_>) -> bool {
+    let mut s = InsertSearcher {
+        cx,
+        receiver: contains_expr.receiver,
+        value: contains_expr.value,
+        should_lint: false,
+    };
+    s.visit_expr(expr);
+    s.should_lint
+}
+
+struct InsertSearcher<'cx, 'tcx> {
+    cx: &'cx LateContext<'tcx>,
+    /// The receiver expression used in the contains call.
+    receiver: &'tcx Expr<'tcx>,
+    /// The value expression used in the contains call.
+    value: &'tcx Expr<'tcx>,
+    /// Whether or a lint shoud be emitted.
+    should_lint: bool,
+}
+
+impl<'tcx> Visitor<'tcx> for InsertSearcher<'_, 'tcx> {
+    fn visit_block(&mut self, block: &'tcx Block<'_>) {
+        for stmt in block.stmts {
+            self.visit_stmt(stmt);
+        }
+        if let Some(expr) = block.expr {
+            self.visit_expr(expr);
+        }
+    }
+
+    fn visit_expr(&mut self, expr: &'tcx Expr<'_>) {
+        match try_parse_insert(self.cx, expr) {
+            Some(insert_expr) => {
+                if SpanlessEq::new(self.cx).eq_expr(self.receiver, insert_expr.receiver)
+                    && SpanlessEq::new(self.cx).eq_expr(self.value, insert_expr.value)
+                {
+                    self.should_lint = true;
+                }
+            },
+            _ => {
+                if let ExprKind::Block(block, _) = expr.kind {
+                    self.visit_block(block);
+                }
+            },
+        }
+    }
+
+    fn visit_stmt(&mut self, stmt: &'tcx Stmt<'_>) {
```

Comment:
This is unnecessary, the default implementation calls `walk_stmt` which does something like this already